Tan11389/nmea location data source

tanneryould opened the pull request:
Adds a sample to demonstrate how to use the NMEA Location Data Source API

ldanzinger commented on ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,49 @@
+    Button {
+        id: button
+        anchors.horizontalCenter: parent.horizontalCenter
+        y: 5
```
couple comments:
* not advisable to use hardcoded x/y values. In this case, you could use `anchors.margins: 5` to achieve the same effect
* I'd opt for the button to be on the bottom. Having at the top is a little distracting and hard to reach with thumb when on a phone

ldanzinger commented on ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.md:
```diff
@@ -0,0 +1,34 @@
+## How it works
+
+1. Load NMEA sentences from a local file.
+2. Parse the NMEA sentence strings, and push data into `NMEALocationDataSource`.
+3. Set the `NMEALocationDataSource` to the location display's data source.
```
```suggestion
2. Parse the NMEA sentence strings, and push data into `NmeaLocationDataSource`.
3. Set the `NmeaLocationDataSource` to the location display's data source.
```

ldanzinger replied:
There are more throughout this readme and the QML version but I won't mark them all

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+            id: nmeaLocationDataSource
+            receiverSpatialReference: SpatialReference { wkid: 4326 }
+        }
+        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter;
```
semicolons only at end of javascript code blocks, not properties

```suggestion
        locationDisplay.autoPanMode: Enums.LocationDisplayAutoPanModeRecenter
```

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+        repeat: true
+        property var iterator: 0
+        onTriggered: {
+
```
```suggestion
```

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+                timer.start();
+                nmeaLocationDataSource.start();
+                mapView.locationDisplay.start();
+            }
+            else {
```
```suggestion
            } else {
```

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+        id: timer
+        interval: 1000
+        repeat: true
+        property var iterator: 0
```
I'd probably call this count or something else with less other meanings.

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+        }
+    }
+
+    property var mockNmeaData: []
```
usually we put global properties at the top of the file for visibility/organization (near line 27)

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+
+    FileFolder {
+        id: mockNmeaDataFile
+        path: ":/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
```
```suggestion
        path: "qrc:/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
```

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+
+    Button {
+        id: button
+        anchors.horizontalCenter: parent.horizontalCenter
+        y: 5
```
same comments as with c++

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+        anchors.horizontalCenter: parent.horizontalCenter
+        y: 5
+        width: 200
+        property bool sampleStarted: false
```
same comment as with C++ - different name suggested

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+    }
+}
```
code looks good in the QML API sample, but could use more comments throughout explaining what is happening

ldanzinger commented on ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/README.metadata.json:
```diff
@@ -0,0 +1,30 @@
+    "redirect_from": "",
+    "relevant_apis": [
+        "LocationDisplay",
+        "NMEALocationDataSource"
```
Same issue in the QML version

tanneryould commented on ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:
```diff
@@ -0,0 +1,142 @@
+  // Load simulated NMEA sentences for sample
+  if (m_mockNmeaSentences.isEmpty())
+  {
+    QString filePath = ":/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea";
```
Weirdly the `qrc:/` prefix breaks it. This runs on an emulated Android as it is

ldanzinger replied:
What does your path look like with the prefix?

tanneryould replied:
`"qrc:/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"`

In past samples, `qrc:/` is used as a prefix when creating a `QUrl` but `QFile::setFileName` expects a `const QString &name`.

tanneryould replied:
`QFile mockDataFile(QUrl("qrc:/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea").toLocalFile());` doesn't work either unfortuantely

ldanzinger replied:
Ok, what you have is good then

tanneryould commented:
Thank you for all of your comments. I believe I've addressed everything and cleaned up a few other areas of sloppy coding. 24 requested changes is certainly a low bar to improve on.

ldanzinger commented on ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.cpp:
```diff
@@ -101,37 +108,40 @@ void DisplayDeviceLocationWithNmeaDataSources::start()
 }
 
-bool DisplayDeviceLocationWithNmeaDataSources::loadMockDataFile(QString filePath)
+bool DisplayDeviceLocationWithNmeaDataSources::loadMockDataFile(const QString &filePath)
```
this is the style we try to use:

```suggestion
bool DisplayDeviceLocationWithNmeaDataSources::loadMockDataFile(const QString& filePath)
```

ldanzinger commented on ArcGISRuntimeSDKQt_CppSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.h:
```diff
@@ -50,20 +50,20 @@ class DisplayDeviceLocationWithNmeaDataSources : public QObject
   Esri::ArcGISRuntime::MapQuickView* mapView() const;
   void setMapView(Esri::ArcGISRuntime::MapQuickView* mapView);
-  bool loadMockDataFile(QString filePath);
+  bool loadMockDataFile(const QString &filePath);
```
```suggestion
  bool loadMockDataFile(const QString& filePath);
```

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -44,69 +47,94 @@ Rectangle {
         id: button
         anchors.horizontalCenter: parent.horizontalCenter
-        y: 5
+        anchors.bottom: parent.bottom
+        anchors.margins: 25
```
anchors can be grouped:

```suggestion
        anchors {
            bottom: parent.bottom
            margins: 25
        }
```

ldanzinger commented on ArcGISRuntimeSDKQt_QMLSamples/Maps/DisplayDeviceLocationWithNmeaDataSources/DisplayDeviceLocationWithNmeaDataSources.qml:
```diff
@@ -44,69 +47,94 @@ Rectangle {
     FileFolder {
         id: mockNmeaDataFile
-        path: ":/Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
+        path: "://Samples/Maps/DisplayDeviceLocationWithNmeaDataSources/redlands.nmea"
```
is this change intentional?

tanneryould replied:
Good catch. It's not. It's left over from testing how I can include `qrc` as we discussed. It still runs, but it should be removed.

ldanzinger reviewed (approved):
Looks good - have a couple more formatting changes suggested, but no need to re-review before merging.
